File: C/82. Remove Duplicates from Sorted List II.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* deleteDuplicates(struct ListNode* head) {
    struct ListNode **walk = &head;
    int dup, flag;

    flag = 0;
    while (*walk) {
        if ((*walk)->next != NULL) {
            dup = (*walk)->val == (*walk)->next->val;
        } else {
            dup = 0;
        }

        if (dup == 1) {
            *walk = (*walk)->next;  /* Delete duplication.  */
            flag = 1;
        } else if (flag == 1) {
            *walk = (*walk)->next;
            flag = 0;
        } else {
            walk = &(*walk)->next;  /* Step forward. */
        }
    }

    return head;
}

File: Python3/82. Remove Duplicates from Sorted List II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        dummy, dummy.next = ListNode(0), head
        prev, cur = dummy, head
        dup = False
        while cur:
            if cur.next and cur.val == cur.next.val:
               dup = True
               cur.next = cur.next.next
            else:
                if dup is True:
                    prev.next, cur = cur.next, cur.next
                    dup = False
                else:
                    prev, cur = cur, cur.next
        return dummy.next